drivers/kscan/kscan_stmpe811.c

```diff
@@ -0,0 +1,440 @@
+			    kscan_callback_t callback)
+{
+	struct stmpe811_data *data = dev->data;
+
+	if (!callback) {
+		LOG_ERR("Invalid callback (NULL)");
+		return -EINVAL;
+	}
+
+	data->callback = callback;
+
+	return 0;
+}
+
+static int stmpe811_init(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+	struct stmpe811_data *data = dev->data;
+
+	int ret = 0;
+	uint16_t chip_id = 0;
+	uint8_t cmd = 0;
+	uint8_t reg = 0;
+
+	data->dev = dev;
+	k_work_init(&data->work, stmpe811_work_handler);
+
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("I2C controller device not ready");
+		return -ENODEV;
+	}
+
+	cmd = STMPE811_REG_CHIP_ID;
+	ret = i2c_burst_read_dt(&config->bus, STMPE811_REG_CHIP_ID, (uint8_t *)&chip_id, 2);
+	if (ret < 0) {
+		LOG_ERR("failed to read Chip ID");
+		return ret;
+	}
+
+	chip_id = sys_be16_to_cpu(chip_id);
+
+	if (chip_id != STMPE811_ID) {
+		LOG_ERR("wrong chip id, returned 0x%x", chip_id);
+		return -ENODEV;
+	}
+
+	ret = stmpe811_reset(dev);
+	if (ret < 0) {
+		LOG_ERR("failed to reset");
+		return ret;
+	}
+
+	/* Set the Functionalities to be Enabled */
+	reg &= ~(STMPE811_TS_FCT | STMPE811_ADC_FCT);
+
+	/* Write the new register value */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_REG_SYS_CTRL2, reg);
+	if (ret < 0) {
+		LOG_ERR("failed to write SYS_CTRL2");
+		return ret;
+	}
+
+	/* Select TSC pins in TSC alternate mode */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_REG_GPIO_ALT_FUNCT, 0x0);
+	if (ret < 0) {
+		LOG_ERR("failed to write ALT_FUNCT");
+		return ret;
+	}
+
+	/* Select Sample Time, bit number and ADC Reference */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_REG_ADC_CTRL1, 0x49);
+	if (ret < 0) {
+		LOG_ERR("failed to write ADC_CTRL1");
+		return ret;
+	}
+
+	k_msleep(2);
+
+	/* Select the ADC clock speed: 3.25 MHz */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_REG_ADC_CTRL2, 0x01);
+	if (ret < 0) {
+		LOG_ERR("failed to write ADC_CTRL2");
+		return ret;
+	}
+
+	/* Configuration:
+	 * Touch average control    : 4 samples
+	 * Touch detect delay       : 1 mS
+	 * Touch delay time         : 500 uS
+	 * Panel driver setting time: 500 uS
+	 */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_REG_TSC_CFG, 0x9a);
+	if (ret < 0) {
+		LOG_ERR("failed to write TSC_CFG");
+		return ret;
+	}
+
+	/* Configure the Touch FIFO threshold: 1 points reading */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_REG_FIFO_TH, 0x01);
+	if (ret < 0) {
+		LOG_ERR("failed to write FIFO_TH");
+		return ret;
+	}
+
+	/* Clear the FIFO memory content. */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_REG_FIFO_STA, 0x01);
+	if (ret < 0) {
+		LOG_ERR("failed to write FIFO_STA");
+		return ret;
+	}
+
+	/* Put the FIFO back into operation mode  */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_REG_FIFO_STA, 0x00);
+	if (ret < 0) {
+		LOG_ERR("failed to write FIFO_STA");
+		return ret;
+	}
+
+	/* Set the driving capability (limit) of the device for TSC pins: 50mA */
```

Comment:
A lot of this initialization seems like it might belong in DT, actually. I will leave it to the ST maintainers to comment further.